### PR TITLE
Add matchers filtering field to rules api

### PR DESCRIPTION
**Pull request by euniceek**
**What this PR does**:
This PR adds support for label matchers as URL query params to the `rules` API. It filters both recording and alerting rules via their labels. 

Prometheus recently merged in a change to support this functionality: https://github.com/prometheus/prometheus/pull/10194

**Checklist**
- [X] Tests updated
- [ ] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by alanprot** [comment]
Thanks!

**Review by alanprot — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -27,6 +27,7 @@
 * [BUGFIX] Querier: Enforce max query length check for `/api/v1/series` API even though `ignoreMaxQueryLength` is set to true. #6018
 * [BUGFIX] Ingester: Fix issue with the minimize token generator where it was not taking in consideration the current ownerhip of an instance when generating extra tokens. #6062
 * [ENHANCEMENT] Ruler: Add support for filtering by `state` and `health` field on Rules API. #6040
+* [ENHANCEMENT] Ruler: Add support for filtering by `match` field on Rules API. #6083
```
Comment (yeya24) [inline CHANGELOG.md:30]: Nit. Can we move this entry and the one above to the `ENHANCEMENT` section? Cortex changelog entries are ordered as `CHANGE, FEATURE, ENHANCEMENT, BUGFIX`. 

https://github.com/cortexproject/cortex/blob/master/RELEASE.md#prepare-your-release

File: pkg/ruler/ruler_test.go [hunk 2]
```diff
@@ -39,6 +39,7 @@ import (
 	"gopkg.in/yaml.v3"
 
 	"github.com/cortexproject/cortex/pkg/cortexpb"
+	github_com_cortexproject_cortex_pkg_cortexpb "github.com/cortexproject/cortex/pkg/cortexpb"
```
Comment (yeya24) [inline pkg/ruler/ruler_test.go:42]: This seems duplicates. Can we just use `cortexpb`?

File: pkg/ruler/ruler.go [hunk 3]
```diff
@@ -1320,3 +1333,40 @@ func returnByState(requestState string, alertState string) bool {
...
+func matches(matcherSets [][]*labels.Matcher, l labels.Labels) bool {
+	if len(matcherSets) == 0 {
+		return true
+	}
+
+	for _, matchers := range matcherSets {
+		for _, m := range matchers {
+			if v := l.Get(m.Name); m.Matches(v) {
+				return true
+			}
```
Comment (yeya24) [inline pkg/ruler/ruler.go:1368]: This seems wrong. I understand that the rule labels need to match all matchers of `matchers` if the label name exists so it is an `AND`.
Only `matcherSets` is an `OR`.

https://github.com/prometheus/prometheus/pull/10194/files#diff-9008036c169df3fe4143c58b08426516bd3fd135f182cacc8b9326541fa55235R167 seems the correct implementation.

Comment (yeya24) [reply pkg/ruler/ruler.go:1368]: Let's also add unit tests to cover this function

File: pkg/ruler/ruler.proto [hunk 4]
```diff
@@ -26,6 +26,7 @@ message RulesRequest {
   string type = 4;
   string state = 5;
   string health = 6;
+  repeated string matches = 7;
```
Comment (yeya24) [inline pkg/ruler/ruler.proto:29]: Let's rename it to `matchers`? `matches` sounds weird.

File: pkg/ruler/ruler.go (hunk 3 above)
Comment (alanprot) [reply pkg/ruler/ruler.go:1368]: good catch!

File: pkg/ruler/ruler.proto (hunk 4 above)
Comment (alanprot) [reply pkg/ruler/ruler.proto:29]: +1

File: pkg/ruler/ruler.go [hunk 5]
```diff
@@ -1320,3 +1333,49 @@ func returnByState(requestState string, alertState string) bool {
...
+// matchesMatcherSets ensures all matches in each matcher set are ANDed and the set of those is ORed.
+func matchesMatcherSets(matcherSets [][]*labels.Matcher, l labels.Labels) bool {
+	if len(matcherSets) == 0 {
+		return true
+	}
+
+	var ok bool
+	for _, matchers := range matcherSets {
+		if matches(l, matchers...) {
+			ok = true
```
Comment (yeya24) [inline pkg/ruler/ruler.go:1377]: It is unnecessary to introduce `ok` here as it keeps checking `matchers` for all `matcherSets`.
We can return early if at least one matches.

**Review by yeya24 — APPROVED** [review]

**Pull request merged**